go/src/net/ConnectionPool.go

```diff
@@ -0,0 +1,247 @@
+	conf "github.com/vesoft-inc/nebula-clients/go/src/conf"
+	data "github.com/vesoft-inc/nebula-clients/go/src/data"
+)
+
+var defaultPoolConfigs = conf.GetDefaultConf()
+
+type RespData struct {
+	Resp *graph.ExecutionResponse
+	Err  error
+}
+
+type ConnectionPool struct {
+	idleConnectionQueue   list.List
+	activeConnectionQueue list.List
+	addresses             []*data.HostAddress
+	conf                  *conf.PoolConfig
+	loadBalancer          *LoadBalancer
+	ConnectionInUse       map[*Connection]bool
+	rwLock                sync.RWMutex
+}
+
+func (pool *ConnectionPool) InitPool(addresses []*data.HostAddress, conf *conf.PoolConfig) error {
+	// Process domain to IP
+	err := data.DomainToIP(addresses)
+	if err != nil {
+		log.Printf("Failed to find IP, error: %s \n", err.Error())
+		return err
+	}
+
+	pool.addresses = addresses
+	pool.conf = conf
+	pool.loadBalancer = NewLoadbalancer(addresses, pool)
+	pool.ConnectionInUse = make(map[*Connection]bool)
+
+	// Check input
+	if len(addresses) == 0 {
+		noAddresErr := errors.New("Failed to reconnect: no avaliable connection in the connection pool")
+		log.Print(noAddresErr)
+		return noAddresErr
+	}
+	if conf == nil {
+		noConfErr := errors.New("Failed to reconnect: no avaliable connection in the connection pool")
+		log.Print(noConfErr)
+		return noConfErr
+	}
+
+	for i := 0; i < pool.conf.MinConnPoolSize; i++ {
+		// Simple round-robin
+		newConn := NewConnection(*pool.addresses[i%len(addresses)])
+		// Not being used
+		pool.ConnectionInUse[newConn] = false
+		// Open connection to host
+		err := newConn.Open(newConn.SeverAddress, *pool.conf)
+		if err != nil {
+			log.Printf("Failed to open connection, error: %s \n", err.Error())
+			return err
+		}
+
+		pool.loadBalancer.IncreaseWorkload(&newConn.SeverAddress)
+		// Mark connection as in use
+		pool.ConnectionInUse[newConn] = false
+		pool.idleConnectionQueue.PushBack(newConn)
+	}
+	return nil
+}
+
+func (pool *ConnectionPool) GetSession(username, password string) (*Session, error) {
+	if pool.conf == nil {
+		notInitialized := errors.New("Failed to get session: There is no config in the connection pool")
+		log.Println(notInitialized)
+		return nil, notInitialized
+	}
+
+	// Get valid and usable connection
+	conn, err := pool.GetIdleConn()
+	if err != nil {
+		log.Printf("Failed to get session, error: %s", err.Error())
+		return nil, err
+	}
+
+	// Authenticate
+	resp, err := conn.Authenticate(username, password)
+	if err != nil || resp.GetErrorCode() != graph.ErrorCode_SUCCEEDED {
+		log.Printf("Failed to authenticate with the given credential, error: %s", err.Error())
+		return nil, err
+	}
+
+	pool.loadBalancer.ValidateHost(&conn.SeverAddress)
+
+	sessionID := resp.GetSessionID()
+	// Create new session
+	newSession := newSession(sessionID, conn, pool)
+
+	pool.rwLock.Lock()
+	// Mark connection as in use
+	pool.ConnectionInUse[conn] = true
+	// Add connction to active queue
+	pool.activeConnectionQueue.PushBack(conn)
```

Comment:
you can move them to `GetIdleConn`, when the session call `GetIdleConn`, the connection is not in you activequeue.